File: Leetcode Easy/1046LastStoneWeight.py
#https://leetcode.com/problems/last-stone-weight/description/
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap = [] #This is a min heap, but is actually max heap because we mulitply by -1
        for stone in stones: 
            heapq.heappush(heap, -1 * stone)
        while len(heap) > 1:
            stone1 = -1 * heapq.heappop(heap) #max
            stone2 = -1 * heapq.heappop(heap) #min
            if stone1 != stone2:
                heapq.heappush(heap, -1*(stone1-stone2))
        if len(heap) == 0: return 0
        else: return -1 * heap[0]

File: C++/Leetcode Easy/1046LastStoneWeight.cpp
#include <queue> 
#include <vector>
#include <cstdlib> 
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        for (int i = 0; i < stones.size(); ++i) pq.push(stones[i]);
        while (pq.size() > 1) {
            int stone1 = pq.top();
            pq.pop();
            int stone2 = pq.top();
            pq.pop();
            if (stone1 == stone2) continue;
            pq.push(abs(stone1-stone2));
        }
        if (pq.size() == 0) return 0;
        else return pq.top();
    }
};

#include <queue>
class Solution {
public:
    priority_queue<int> maxHeap;
    int lastStoneWeight(vector<int>& stones) {
        for (int stone : stones){
            maxHeap.push(stone);
        }
        while (maxHeap.size() > 1){
            int stone1 = maxHeap.top();
            maxHeap.pop();
            int stone2 = maxHeap.top();
            maxHeap.pop();
            maxHeap.push(stone1-stone2);
        }
        return maxHeap.top();
    }
};